[docs] Add integration, base concepts (#4080)
mui/toolpad: docs/data/toolpad/core/introduction/base-concepts.md

---
title: Toolpad Core - Base Concepts
---

# Base concepts

<p class="description">Understand the fundamental concepts of Toolpad Core to effectively integrate and use it in your projects.</p>

## Imports

Toolpad Core components can be imported directly from the `@toolpad/core` package. This allows you to use them alongside your existing Material UI or other components.

```tsx
import Button from '@mui/material/Button';
import { DashboardLayout } from '@toolpad/core/DashboardLayout';
```

## Component Hierarchy

> The Toolpad Core library is designed to work within your existing React application structure. The key component in this hierarchy is the `AppProvider`.

Reviewer: ```suggestion
The Toolpad Core library is designed to work under different React runtimes such as Next.js, vite, or even your custom setup. Many of its components rely on specific functionality of the runtime it's used under. For example: routing. The key component in making the components runtime aware is the `AppProvider` component.
```

    <AppProvider>
      <Component {...pageProps} />
    </AppProvider>
  );
}
```

By wrapping your application with `AppProvider`, you ensure that all other Toolpad components you use have access to the necessary context and functionality.

The `AppProvider` component accepts props to configure the app's navigation, theme, branding, router, authentication, and session, like so:

```tsx
<AppProvider
  navigation={NAVIGATION}
  theme={theme}
  branding={BRANDING}
  router={router}
  authentication={AUTHENTICATION}
  session={session}
>
  {props.children}
</AppProvider>
```

Head over to the [AppProvider](/toolpad/core/react-app-provider/) page for more details and examples of the usage of all props.

:::info
Toolpad Core doesn't handle routing itself. Instead, it's designed to integrate seamlessly with your existing routing solution, whether you're using:

- Next.js App Router
- Next.js Pages Router
- React Router
- Or any other routing library which implements the same interface

You can pass the router implementation to the `AppProvider` component using the `router` prop. You do not need to pass this prop if you are using Next.js, since it has a file-system based router.

:::

> 

Reviewer: Maybe we can add that for Next.js you don't need to set this up yourself. We have an `AppProvider` exported from `@toolpad/core/nextjs` that already sets it up correctly.

Author: I could add it as a separate warning from this info callout and draw specific attention to the `@toolpad/core/nextjs` export

---
title: Toolpad Core - Base Concepts
---

> # Base Concepts

Reviewer 2: ```suggestion
# Base concepts
```

> <p class="description">Understanding the fundamental concepts of Toolpad Core to effectively integrate and use it in your projects.</p>

Reviewer 2: ```suggestion
<p class="description">Understand the fundamental concepts of Toolpad Core to effectively integrate and use it in your projects.</p>
```

---
title: Toolpad Core - Base Concepts
---

# Base Concepts

<p class="description">Understanding the fundamental concepts of Toolpad Core to effectively integrate and use it in your projects.</p>

## Imports

> Toolpad Core components can be imported directly from the `@toolpad/core` package. This allows you to use them alongside your existing Material UI or other components.

Reviewer: ```suggestion
Toolpad Core components can be imported directly from the `@toolpad/core` package. This allows you to use them alongside your existing Material UI or other components.
```

```tsx
import { Button } from '@mui/material';
import { DashboardLayout } from '@toolpad/core/DashboardLayout';
```

## Component Hierarchy

The Toolpad Core library is designed to work within your existing React application structure. The key component in this hierarchy is the `AppProvider`.

### App Provider

The `AppProvider` acts as a bridge between your application's runtime and Toolpad components. It should wrap your entire application or the part of your application where you want to use Toolpad components.

```tsx
import { AppProvider } from '@toolpad/core/AppProvider';

function MyApp({ Component, pageProps }) {
  return (
    <AppProvider>
      <Component {...pageProps} />
    </AppProvider>
  );
}
```

By wrapping your application with `AppProvider`, you ensure that all other Toolpad components you use have access to the necessary context and functionality.

> #### AppProvider Props

Reviewer 3: Oh I see they're actually covered more in the `DashboardLayout`.
But maybe they could be described in the AppProvider documentation and we could link to there from here?
And keep this part in this page more generic and less specific to implementation?

> #### AppProvider Props

Author: We could do that to reduce maintenance load. The intention behind me doing this here is considering that this page could be something that a prospective user reads first before installing any components - so to make this document a "pitch", it should be the smoothest experience possible, which to me means fewer links, more inline demos. Let me know if that makes sense to you.

> #### AppProvider Props

Reviewer 3: I just think we can be a bit more generic here and not cover every `AppProvider` prop as extensively which is a bit too much information also... So just mentioning the most important things here: probably routing, theming, authentication... in a more general way. We can have code examples too, just maybe condense and stick to the most important features as opposed to listing and describing every prop.

But it's also something we can discuss and do later if it's not super consensual.

> #### AppProvider Props

Author: Agree with covering it more generally and not overloading with information - you're right! I'll change this a bit

> #### AppProvider Props

Reviewer 2: I agree with Pedro, I would even remove the one line explainers for each prop and directly show a code demo like below. It should give users the basic idea. 

```
<AppProvider
              navigation={NAVIGATION}
              branding={BRANDING}
              session={session}
              theme={theme}
              authentication={AUTHENTICATION}
            >
              {props.children}
</AppProvider>
```

The `AppProvider` component accepts the following props:

> - `theme`: The theme to be used by the app in light/dark mode. If you are using [Material UI with a custom theme](https://mui.com/material-ui/customization/theming/), you can directly pass it here.

Reviewer: ```suggestion
- `theme`: The theme to be used by the app in light/dark mode. If you are using [Material UI with a custom theme](https://mui.com/material-ui/customization/theming/), you can directly pass it here.
```
